modules/standard/Math.chpl

```diff
@@ -1232,4 +1232,64 @@ module Math {
     extern proc yn(n: c_int, x: real(64)): real(64);
     return yn(n.safeCast(c_int), x);
   }
+
+  @chpldoc.nodoc
+  private inline proc fmaImplExternCall(x: real(?w), y: real(w), z: real(w)) {
+    extern proc fmaf(x: real(32), y: real(32), z: real(32)): real(32);
+    extern proc fma(x: real(64), y: real(64), z: real(64)): real(64);
+    var ret: x.type;
+    select x.type {
+      when real(32) do ret = fmaf(x, y, z);
+      when real(64) do ret = fma(x, y, z);
+      otherwise do compilerError('Unhandled type in fma() impl!');
+    }
+    return ret;
+  }
+
+  @chpldoc.nodoc
+  private inline proc fmaSelectPrimitiveOrExternCall(x: ?t, y: t, z: t) {
+    import ChplConfig;
+    param isLlvmCompile: bool = (ChplConfig.CHPL_TARGET_COMPILER == "llvm");
+
+    compilerAssert(t == real(32) || t == real(64));
+
+    // The backend will emit a 'llvm.fma.*' instruction, which should be
+    // optimized into a hardware instruction if the architecture is
+    // specified and `--specialize` is thrown.
+    if isLlvmCompile {
+      return __primitive("fma", x, y, z);
+
+    // Easier to call C 'fma()' here than have the backend do it. It's up
+    // to the C compiler if any optimization occurs here at all. We have
+    // to call 'fma()' for correctness, as (x*y+z) will introduce more error.
+    // TODO: Is there a path for the builtin Clang to reliably emit a
+    // hardware instruction, e.g., through a Clang intrinsic call?
+    } else {
+      return fmaImplExternCall(x, y, z);
+    }
+  }
+
+  /** Performs a fused multiply-add operation that multiplies ``x`` and ``y``
+      and adds ``z`` to the result. The advantage of ``fma()`` over the
+      expression ``(x*y)+z`` is that it avoids the additional error
+      introduced by performing two separate floating point operations.
```

Comment:
Well, in some settings, it will be faster, since it's one floating point instruction rather than two.

Reply:
I added a sentence about it possibly being faster.